--- a/setup.orig.py	2020-07-15 17:54:25.805747151 +0200
+++ b/setup.py	2020-07-15 17:53:15.841073182 +0200
@@ -1510,7 +1510,7 @@
     if is_Fedora() or is_CentOS() or is_RedHat() or FREEBSD:
         libexec = "libexec"
     else:
-        libexec = "lib"
+        libexec = "libexec"
     if LINUX or FREEBSD:
         if scripts_ENABLED:
             scripts += ["fs/bin/xpra_udev_product_version", "fs/bin/xpra_signal_listener"]
