--- a/setup.orig.py	2020-06-05 11:40:10.000000000 +0200
+++ b/setup.py	2020-07-15 11:07:19.084531962 +0200
@@ -1899,8 +1899,8 @@
     #cairo workaround:
     add_cython_ext("xpra.client.gtk3.cairo_workaround",
                 ["xpra/client/gtk3/cairo_workaround.pyx"],
-                **pkgconfig("py3cairo")
-                )
+                ["@@python3_sitearch@@"]
+                )
 
 if client_ENABLED or server_ENABLED:
     add_packages("xpra.codecs.argb")
--- a/xpra/client/gtk3/cairo_workaround.orig.pyx	2020-05-10 19:00:53.000000000 +0200
+++ b/xpra/client/gtk3/cairo_workaround.pyx	2020-07-15 11:17:57.586807490 +0200
@@ -35,7 +35,7 @@
         pass
     void * PyCapsule_Import(const char *name, int no_block)
 
-cdef extern from "cairo/cairo.h":
+cdef extern from "cairo/include/py3cairo.h":
     ctypedef struct cairo_surface_t:
         pass
 
@@ -59,7 +59,7 @@
     void cairo_surface_flush (cairo_surface_t *surface)
     void cairo_surface_mark_dirty (cairo_surface_t *surface)
 
-cdef extern from "pycairo/py3cairo.h":
+cdef extern from "cairo/include/py3cairo.h":
     ctypedef struct Pycairo_CAPI_t:
         pass
     ctypedef struct PycairoSurface:
