diff -up xpra-1.0/xpra/codecs/webp/encode.pyx.orig xpra-1.0/xpra/codecs/webp/encode.pyx
--- xpra-1.0/xpra/codecs/webp/encode.pyx.orig	2016-12-06 10:56:27.000000000 +0000
+++ xpra-1.0/xpra/codecs/webp/encode.pyx	2016-12-10 11:13:14.305360914 +0000
@@ -8,7 +8,7 @@ import os
 from xpra.log import Logger
 log = Logger("encoder", "webp")
 
-from xpra.util envbool
+from xpra.util import envbool
 cdef int LOG_CONFIG = envbool("XPRA_WEBP_LOG_CONFIG", False)
 cdef int USE_THREADS = envbool("XPRA_WEBP_THREADING", True)
 
