Index: /tags/v0.17.x/src/setup.py
===================================================================
--- /tags/v0.17.x/src/setup.py	(revision 12943)
+++ /tags/v0.17.x/src/setup.py	(revision 12944)
@@ -2312,4 +2312,5 @@
 if dec_avcodec2_ENABLED:
     avcodec2_pkgconfig = pkgconfig("avcodec", "avutil")
+    add_to_keywords(avcodec2_pkgconfig, 'extra_compile_args', "-Wno-error=deprecated-declarations")
     cython_add(Extension("xpra.codecs.dec_avcodec2.decoder",
                 ["xpra/codecs/dec_avcodec2/decoder.pyx"]+membuffers_c,
