--- setup.orig.py	2017-02-01 06:52:14.000000000 +0100
+++ setup.py	2017-04-21 21:36:02.901697592 +0200
@@ -184,7 +184,7 @@
 enc_proxy_ENABLED       = DEFAULT
 enc_x264_ENABLED        = DEFAULT and pkg_config_ok("--exists", "x264", fallback=WIN32)
 enc_x265_ENABLED        = DEFAULT and pkg_config_ok("--exists", "x265")
-enc_xvid_ENABLED        = DEFAULT and pkg_config_ok("--exists", "xvid")
+enc_xvid_ENABLED        = DEFAULT and os.path.exists("/usr/include/xvid.h")
 pillow_ENABLED          = DEFAULT
 webp_ENABLED            = False
 vpx_ENABLED             = DEFAULT and pkg_config_version("1.3", "vpx", fallback=WIN32)
@@ -2353,7 +2353,7 @@
 
 toggle_packages(enc_xvid_ENABLED, "xpra.codecs.enc_xvid")
 if enc_xvid_ENABLED:
-    xvid_pkgconfig = pkgconfig("xvid")
+    xvid_pkgconfig = pkgconfig()
     cython_add(Extension("xpra.codecs.enc_xvid.encoder",
                 ["xpra/codecs/enc_xvid/encoder.pyx", buffers_c],
                 **xvid_pkgconfig))
